### ISSUE TITLE
Jupyter notebook / IPython Magic for eralchemy

### ISSUE THREAD

**Issue by psychemedia** [Alexis-benoist/eralchemy#45]
I made a start on some IPython magic for rendering `eralchemy` generated images in a Jupyter notebook: [ipython_magic_eralchemy](https://github.com/innovationOUtside/ipython_magic_eralchemy)

It really needs factoring into `eralchemy` and perhaps `ipython-sql` somehow so that it could make use of a persistent SQL connection. ([`ipython-sql` magic](https://github.com/catherinedevlin/ipython-sql/blob/master/src/sql/magic.py) uses `import sql.connection` from that package to persist connections in a notebook so you only have to declare/open them once.)

**Comment by kasium** [comment]
@maurerle can you please assign me to this issue?
@psychemedia I'm sorry for the very late response. I don't fully get your requirement. Can you please explain what you need in eralchemy?

**Comment by psychemedia** [comment]
It was something along the lines of `eralchemy` baxcking something like [`ipython_magic_sqlalchemy_schemadisplay`](https://github.com/innovationOUtside/ipython_magic_sqlalchemy_schemadisplay)

<img width="765" alt="image" src="https://github.com/user-attachments/assets/274c72b4-4f4e-47e3-b9cb-5418b4394545">}

**Comment by kasium** [comment]
So, eralchemy supports that you provide a sql connection string. Based on this a ER diagram is created

**Comment by psychemedia** [comment]
Yes, though more generally: allow the magic to use or create an `eralchemy` connection to a database and then visualise the database structure.

**Comment by maurerle** [comment]
The idea is to have this work in a jupyter notebook which renders the output inline in the notebook.

I am not to familiar with the ipython magic, but we might want to integrate parts of https://github.com/innovationOUtside/ipython_magic_eralchemy/blob/master/eralchemy_magic/eralchemy.py into eralchemy.
Though we should do this in a way which does not make ipython a hard dependency of it

**Comment by kasium** [comment]
@maurerle thanks for the hint. I checked all the code and basically the following is needed if we want to support everything

* make render_er return something which supports the [repr methods of ipython](https://ipython.readthedocs.io/en/stable/config/integrating.html#custom-methods)
* add an iphython/jupyter integration by exposing a magic method and a CLI interface
* support reuse of db connections like in [ipython-sql](https://github.com/catherinedevlin/ipython-sql)

This is all does not come without complexity. Especially we all are no experts and testing would be a hell lot of fun. I would recommend to change `render_er` so that it returns some data. Maybe it could even define the ipython formatter methods. However, I would recommend to leave all other ipython/jupyter stuff aside. I would see this in another package so that eralchemy can focus on its domain

**Comment by psychemedia** [comment]
Here's an example of hacked magic for using eralchemy to generate schema diagrams in Jupyter notebooks and VSCode: https://github.com/innovationOUtside/erlachemy_schemadisplay_anywidget_magic

It also uses a wasm graphviz bundle to render the dot file, packaged using the anywidget framework (i.e. doesn't require an O/S install of graphviz). This works in browser/Chromium/electron based UIs (JupyterLab and notebooks, VS Code, etc and (if you're careful waiting for things to finish), in JupyterLite and marimo (i.e. pyodide based envts that run purely in the browser)).
